beacon-chain/blockchain/process_attestation_helpers.go

```diff
@@ -18,17 +18,48 @@ import (
 	"github.com/prysmaticlabs/prysm/v5/time/slots"
 )
 
-// getAttPreState retrieves the att pre state by either from the cache or the DB.
-func (s *Service) getAttPreState(ctx context.Context, c *ethpb.Checkpoint) (state.ReadOnlyBeaconState, error) {
-	// If the attestation is recent and canonical we can use the head state to compute the shuffling.
+func (s *Service) getRecentPreState(ctx context.Context, c *ethpb.Checkpoint) state.ReadOnlyBeaconState {
 	headEpoch := slots.ToEpoch(s.HeadSlot())
+	if c.Epoch < headEpoch {
+		return nil
+	}
+	if !s.cfg.ForkChoiceStore.IsCanonical([32]byte(c.Root)) {
+		return nil
+	}
 	if c.Epoch == headEpoch {
 		targetSlot, err := s.cfg.ForkChoiceStore.Slot([32]byte(c.Root))
-		if err == nil && slots.ToEpoch(targetSlot)+1 >= headEpoch {
-			if s.cfg.ForkChoiceStore.IsCanonical([32]byte(c.Root)) {
-				return s.HeadStateReadOnly(ctx)
-			}
+		if err != nil {
+			return nil
 		}
+		if slots.ToEpoch(targetSlot)+1 < headEpoch {
+			return nil
+		}
+		st, err := s.HeadStateReadOnly(ctx)
+		if err != nil {
+			return nil
+		}
+		return st
+	}
+	slot, err := slots.EpochStart(c.Epoch)
+	if err != nil {
+		return nil
+	}
+	st, err := s.HeadState(ctx)
+	if err != nil {
+		return nil
+	}
+	st, err = transition.ProcessSlotsUsingNextSlotCache(ctx, st, c.Root, slot)
+	if err != nil {
+		return nil
+	}
```

Comment:
Why wouldn't we want to cache the result of this in the check point attestation cache? Seems to me if there are N attestations trigger this condition then we will need to do this N times which could become a dos vector. Are state copy and process slots here that cheap?

Reply:
if the NSC cache is a hit wouldn't this be already in the checkpoint cache? hmm probably better to follow on Slack

Reply:
You're right, we're getting these messages on attestations up to slot 5, so they are not in the NSC, I updated the checkpoint cache here.